sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java

```diff
@@ -0,0 +1,124 @@
+package org.apache.beam.sdk.io.kinesis;
+
+import com.google.common.base.Function;
+import com.google.common.collect.Maps;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Internal shard iterators pool.
+ * It maintains the thread pool for reading Kinesis shards in separate threads.
+ * Read records are stored in a blocking queue of limited capacity.
+ */
+class ShardReadersPool {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ShardReadersPool.class);
+  private static final int DEFAULT_CAPACITY_PER_SHARD = 10_000;
+  private ExecutorService executorService;
+  private BlockingQueue<KinesisRecord> recordsQueue;
+  private Map<String, ShardRecordsIterator> shardIteratorsMap;
+  private final int queueCapacity;
+  private volatile boolean poolOpened = true;
+
+  ShardReadersPool(List<ShardRecordsIterator> iterators) {
+    this(iterators, DEFAULT_CAPACITY_PER_SHARD);
+  }
+
+  ShardReadersPool(List<ShardRecordsIterator> iterators, int queueCapacity) {
+    this.shardIteratorsMap = Maps
+        .uniqueIndex(iterators, new Function<ShardRecordsIterator, String>() {
+
+          @Override
+          public String apply(ShardRecordsIterator input) {
+            return input.getShardId();
+          }
+        });
+    this.queueCapacity = queueCapacity;
+  }
+
+  void start() {
+    executorService = Executors.newFixedThreadPool(shardIteratorsMap.size());
+    recordsQueue = new LinkedBlockingQueue<>(queueCapacity * shardIteratorsMap.size());
+    for (final ShardRecordsIterator shardRecordsIterator : shardIteratorsMap.values()) {
+      executorService.submit(new Runnable() {
+
+        @Override
+        public void run() {
+          readLoop(shardRecordsIterator);
+        }
+      });
+    }
+  }
+
+  private void readLoop(ShardRecordsIterator shardRecordsIterator) {
+    while (poolOpened) {
+      try {
+        List<KinesisRecord> kinesisRecords = shardRecordsIterator.readNextBatch();
+        for (KinesisRecord kinesisRecord : kinesisRecords) {
+          recordsQueue.put(kinesisRecord);
+        }
+      } catch (TransientKinesisException e) {
+        LOG.warn("Transient exception occurred.", e);
+      } catch (InterruptedException e) {
+        LOG.warn("Thread was interrupted, finishing the read loop", e);
+        break;
+      } catch (Throwable e) {
+        LOG.error("Unexpected exception occurred", e);
```

Comment:
You cannot simply swallow the exception here as it may be fatal exception - for example we don't have permissions to read from Kinesis shards. You need to find a way to propagate that exception up.

Reply:
I thought about this issue and I'm not sure what we can do about fatal exceptions at this place. I chose to log them so at least there's a trace of what happened. Stopping the read loop entirely does not seem like a perfect solution either because it still won't stop the KinesisReader from trying to advance. In fact the only unexpected exceptions at this place could be thrown from the Kinesis calls so probably the worst thing that could happen is the permissions loss.

I don't know if other connectors have similar considerations. I've checked KafkaIO sources and haven't seen any special exceptions handling in a read loop.

Reply:
Hmm... it'd be good if someone from Beam maintainers would look at that.